tests/framework/framework/framework.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file was originally copied from https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go
+
+package framework
+
+import (
+	"fmt"
+	"math/rand"
+
+	"kubevirt.io/kubevirt/tests"
+	"kubevirt.io/kubevirt/tests/util"
+
+	"kubevirt.io/client-go/kubecli"
+
+	"k8s.io/client-go/rest"
+
+	. "github.com/onsi/ginkgo"
+)
+
+// Framework supports common operations used by e2e tests; it will keep a client & clean the cluster for you.
+type Framework struct {
+	BaseName string
+
+	UniqueName string
+
+	clientConfig   *rest.Config
+	KubevirtClient kubecli.KubevirtClient
+	// afterEaches is a map of name to function to be called after each test.  These are not
+	// cleared.  The call order is randomized so that no dependencies can grow between
+	// the various afterEaches
+	afterEaches map[string]AfterEachActionFunc
+}
+
+// AfterEachActionFunc is a function that can be called after each test
+type AfterEachActionFunc func(f *Framework, failed bool)
+
+// NewDefaultFramework makes a new framework and sets up a BeforeEach/AfterEach for
+// you (you can write additional before/after each functions).
+func NewDefaultFramework(baseName string) *Framework {
+	return NewFramework(baseName, nil, nil)
+}
+
+// NewFramework creates a test framework.
+func NewFramework(baseName string, client kubecli.KubevirtClient, config *rest.Config) *Framework {
+	f := &Framework{
+		BaseName:       baseName,
+		KubevirtClient: client,
+		clientConfig:   config,
+	}
+	BeforeEach(f.BeforeEach)
+	AfterEach(f.AfterEach)
+
+	return f
+}
+
+// BeforeEach gets a client and clean the cluster.
+func (f *Framework) BeforeEach() {
+	if f.KubevirtClient == nil {
+		By("Creating a kubevirt client")
+		client, err := kubecli.GetKubevirtClient()
+		util.PanicOnError(err)
+		f.KubevirtClient = client
+		config, err := kubecli.GetKubevirtClientConfig()
+		util.PanicOnError(err)
+		f.clientConfig = config
+	}
+	f.UniqueName = fmt.Sprintf("%s-%08x", f.BaseName, rand.Int31())
+	tests.BeforeTestCleanup()
```

Comment:
Sorry I don't understand. This function call makes the cluster clean before all tests, instead of declare it in each single test file. Why do we have to remove it?

Reply:
Is there a way to design it so failed test reporting (with log dump) runs after tests, and cleanup is done later, after that? I suppose this should work, because we use `JustAfterEach` for reporting and we can use `AfterEach` for cleanup. 

Is there a problem that forces us to clean before test? Is it better to clean before?

Reply:
I have started reading a discussion on kubevirt-dev. I can see it mentioned as "an intermediate solution". So maybe I need to finish reading the discussion before reviewing this PR.

Reply:
@brybacki As you said we collect reporting data in the `JustAfterEach` so yes, basically we can cleanup in the `AfterEach`.
Right, the `Framework` was intended, at least by me, as an intermediate solution that should be extended step by step introducing functionalities. But it's ok for me to drop it from this PR and discuss which functionalities we want in the proposal design.